src/vm/analysis/type_checker/natives/iterables.rs

```diff
@@ -0,0 +1,209 @@
+    }
+
+    Ok(argument_type)
+}
+
+pub fn check_special_fold(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(3, args)?;
+    
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::NonFunctionApplication)?;
+    // we will only lookup native or defined functions here.
+    //   you _cannot_ fold a special function.
+    let function_type = get_simple_native_or_user_define(function_name, checker)?;
+    
+    let argument_type = checker.type_check(&args[1], context)?;
+
+    let input_type = match argument_type {
+        TypeSignature::ListType(list_data) => Ok(list_data.destruct().0),
+        TypeSignature::BufferType(_) => Ok(TypeSignature::min_buffer()),
+        _ => Err(CheckErrors::ExpectedListOrBuffer(argument_type))
+    }?;
+
+    let initial_value_type = checker.type_check(&args[2], context)?;
+
+    // fold: f(A, B) -> A
+    //     where A = initial_value_type
+    //           B = list items type
+    
+    // f must accept the initial value and the list items type
+    let return_type = function_type.check_args(&[input_type.clone(), initial_value_type])?;
+
+    // f must _also_ accepts its own return type!
+    let return_type = function_type.check_args(&[input_type, return_type])?;
+    
+    Ok(return_type)
+}
+
+pub fn check_special_concat(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(2, args)?;
+    
+    let lhs_type = checker.type_check(&args[0], context)?;
+    match lhs_type {
+        TypeSignature::ListType(lhs_list) => {
+            let rhs_type = checker.type_check(&args[1], context)?;
+            if let TypeSignature::ListType(rhs_list) = rhs_type {
+                let (lhs_entry_type, lhs_max_len) = lhs_list.destruct();
+                let (rhs_entry_type, rhs_max_len) = rhs_list.destruct();
+
+                if lhs_entry_type.admits_type(&rhs_entry_type) {
+                    let new_len = lhs_max_len.checked_add(rhs_max_len)
+                        .ok_or(CheckErrors::MaxLengthOverflow)?;
+                    let return_type = TypeSignature::list_of(lhs_entry_type, new_len)?;
+                    return Ok(return_type);
+                } else {
+                    return Err(CheckErrors::TypeError(lhs_entry_type, rhs_entry_type).into());
+                }
+            } else {
+                return Err(CheckErrors::TypeError(rhs_type.clone(), TypeSignature::ListType(lhs_list)).into());
+            }
+        },
+        TypeSignature::BufferType(lhs_buff_len) => {
+            let rhs_type = checker.type_check(&args[1], context)?;
+            if let TypeSignature::BufferType(rhs_buff_len) = rhs_type {
+                let size: u32 = u32::from(lhs_buff_len).checked_add(u32::from(rhs_buff_len))
+                    .ok_or(CheckErrors::MaxLengthOverflow)?;
+                let return_type = TypeSignature::buffer_of_size(size);
```

Comment:
`buffer_of_size` looks like it panics if the supplied size is greater than max buffer size, which isn't a good behavior here. This function should instead through a CheckErrors::ValueTooLarge error (which is the error that `TypeSignature::BufferType(size.try_into()?)` would raise.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1139/commits/091b62538e90cf6525faa26ff2bbe7f01c9fe8cf